### PR TITLE
README for Shunting Yard

### PR DESCRIPTION
**Pull request by abhimanyugupta07**
First attempt at an exhaustive README for Shunting Yard.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,61 +1,147 @@
 # Shunting Yard
 
-A Spring Boot app that reads serialized Hive MetaStore Events and builds a YAML file with the information provided in the event which is then passed to [Circus Train](https://github.com/HotelsDotCom/circus-train) to perform the replication.
+Shunting Yard reads serialized Hive MetaStore Events from a queue (currently supports [AWS SQS](https://aws.amazon.com/sqs/)) and replicates the data between two Data lakes. It does this by building a YAML file with the information provided in the event which is then passed to [Circus Train](https://github.com/HotelsDotCom/circus-train) to perform the replication.
```
Comment (AnanaMJ) [inline README.md]: "between two Data lakes": does "data" have to be Uppercase?

File: README.md [hunk 2]
```diff
@@ -1,61 +1,147 @@
...
 
-On the source cluster, copy the file respective emitter JAR to _/usr/lib/hive/lib/_ and configure _/etc/hive/conf/hive-site.xml_ with the properties required by the listener to talk to the messaging infrastructure:
+         export HCAT_LIB=/usr/lib/hive-hcatalog/share/hcatalog/
+         export HIVE_LIB=/usr/lib/hive/lib/
 
-#### SQS
-    hive.metastore.event.listeners = com.hotels.shunting.yard.event.emitter.sqs.listener.SqsMetaStoreEventListener
-    com.hotels.shunting.yard.event.emitter.sqs.queue = https://sqs.<region>.amazonaws.com/<account-id>/<topic-name>-queue.fifo
-    com.hotels.shunting.yard.event.emitter.sqs.group.id = <group-id>
+Note that the paths above are correct as of when this document was last updated but may differ across EMR versions, refer to the [EMR release guide](http://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-release-components.html) for more up to date information if necessary.
```
Comment (AnanaMJ) [inline README.md]: There should be a full stop instead of comma, as there seem to be two separate sentences in the same sentence.

File: README.md [hunk 3]
```diff
@@ -1,61 +1,147 @@
...
+
+    table-replications:
+      - source-table:
+          database-name: source_database
+          table-name: test_table
+        replica-table:
+          database-name: replica_database
+          table-name: test_table_1    
+          
+#### Only Change the target database but the table name remains same as source
```
Comment (AnanaMJ) [inline README.md]: "change" with lowercase "c"

Comment (AnanaMJ) [reply README.md]: And maybe it would be good to rephrase this as "Change only the target database while keeping the same source table name" to keep the same tone of the writing (as an instruction)

File: README.md [hunk 4]
```diff
@@ -1,61 +1,147 @@
...
+
+    table-replications:
+      - source-table:
+          database-name: source_database
+          table-name: test_table
+        replica-table:
+          database-name: replica_database 
+
+#### Only Change the target table name but the database remains same as source
```
Comment (AnanaMJ) [inline README.md]: Same thing here, but the database and table switched around :)

File: README.md [hunk 5]
```diff
@@ -1,61 +1,147 @@
...
+### Shunting Yard configuration reference
+The table below describes all the available configuration values for Shunting Yard.
+
+|Property|Required|Description|
+|:----|:----:|:----|
+|`source-catalog.name`|Yes|A name for the source catalog for events and logging.|
+|`source-catalog.hive-metastore-uris`|No|Fully qualified URI of the source cluster's Hive metastore Thrift service.|
+|`replica-catalog.name`|Yes|A name for the replica catalog for events and logging.|
+|`replica-catalog.hive-metastore-uris`|Yes|Fully qualified URI of the replica cluster's Hive metastore Thrift service.|
+|`event-receiver.configuration-properties.com.hotels.shunting.yard.event.receiver.sqs.queue`|Yes|Fully qualified URI of the [AWS SQS](https://aws.amazon.com/sqs/) Queue to read the hive events from.|
```
Comment (AnanaMJ) [inline README.md]: Uppercase H on "hive"

File: README.md [hunk 6]
```diff
@@ -1,61 +1,147 @@
...
+|`event-receiver.configuration-properties.com.hotels.shunting.yard.event.receiver.sqs.wait.time.seconds`|No|Wait time in seconds for which the receiver will poll the SQS queue for a batch of messages. Default is 10 seconds. Read more about long polling with AWS SQS [here](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-long-polling.html)|
```
Comment (AnanaMJ) [inline README.md]: Full stop at the end 🙂

File: README.md [hunk 7]
```diff
@@ -1,61 +1,147 @@
...
+|`source-table-filter.table-names`|No|A list of tables selected for Shunting Yard replication. Supported Format:`database_1.table_1, database_2.table_2`|
```
Comment (AnanaMJ) [inline README.md]: space after colon:
"format: `database_1.table_1, database_2.table_2`"

and lowercase f on "Format"

File: README.md [hunk 8]
```diff
@@ -1,61 +1,147 @@
...
+|`table-replications[n].source-table.database-name`|No|The name of the database in which the table you wish to replicate is located.|
+|`table-replications[n].source-table.table-name`|No|The name of the table which you wish to replicate.|
+|`table-replications[n].replica-table.database-name`|No|The name of the destination database in which to replicate the table. Defaults to source database name.|
+|`table-replications[n].replica-table.table-name`|No|The name of the table at the destination. Defaults to source table name.|
+
+### Configuring Graphite Metrics
```
Comment (AnanaMJ) [inline README.md]: I think "Metrics" should be with lowercase m. "System architecture" has 2 hashes (compared to 3 here) and doesn't have uppercase for each letter in the subtitle, so this shouldn't either.

**Comment by abhimanyugupta07** [comment]
> I definitely wasn't looking for a missing full stop on purpose :)

I think you were 🙈

File: README.md (hunk 8 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 7 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 6 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 5 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 4 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 3 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 2 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 1 above)
Comment (abhimanyugupta07) [reply README.md]: done.

File: README.md (hunk 3 above)
Comment (AnanaMJ) [reply README.md]: what about the second half?

File: README.md (hunk 7 above)
Comment (AnanaMJ) [reply README.md]: There's still no space after colon

File: README.md (hunk 8 above)
Comment (AnanaMJ) [reply README.md]: Graphite can stay with uppercase, because it's the name of the tool, right?

File: README.md [hunk 9]
```diff
@@ -1,6 +1,6 @@
 # Shunting Yard
 
-Shunting Yard reads serialized Hive MetaStore Events from a queue (currently supports [AWS SQS](https://aws.amazon.com/sqs/)) and replicates the data between two Data lakes. It does this by building a YAML file with the information provided in the event which is then passed to [Circus Train](https://github.com/HotelsDotCom/circus-train) to perform the replication.
+Shunting Yard reads serialized Hive MetaStore Events from a queue (currently supports [AWS SQS](https://aws.amazon.com/sqs/)) and replicates the data between two data lakes. It does this by building a YAML file with the information provided in the event which is then passed to [Circus Train](https://github.com/HotelsDotCom/circus-train) to perform the replication.
```
Comment (massdosage) [inline README.md]: currently supports [AWS SQS](https://aws.amazon.com/sqs/) -> [AWS SQS](https://aws.amazon.com/sqs/) is currently supported

File: README.md [hunk 10]
```diff
@@ -1,61 +1,147 @@
...
 
 ## System architecture
 
 ![Shunting Yard system diagram.](shunting-yard.png "Selected tables getting replicated by Shunting Yard based on Hive events.")
 
-## Start using
+## Install
 
-You can obtain Shunting Yard from Maven Central:
+1. Download the version to use from [Maven Central](https://mvnrepository.com/artifact/com.hotels/shunting-yard-binary) and uncompress it in a directory of your choosing.
```
Comment (massdosage) [inline README.md]: It's a bit unclear here which file one should actually download, for a new user we really need to spell it out for them. I know this is what it says in the Circus Train docs but that could also be improved (once we get the wording sorted out here let's do the same for Circus Train). I think Waggle Dance is a better example of this: https://github.com/HotelsDotCom/waggle-dance/#install

File: README.md [hunk 11]
```diff
@@ -1,61 +1,147 @@
...
 
-[![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.hotels/shunting-yard/badge.svg?subject=com.hotels:shunting-yard)](https://maven-badges.herokuapp.com/maven-central/com.hotels/shunting-yard) [![Build Status](https://travis-ci.org/HotelsDotCom/shunting-yard.svg?branch=master)](https://travis-ci.org/HotelsDotCom/shunting-yard) [![Coverage Status](https://coveralls.io/repos/github/HotelsDotCom/shunting-yard/badge.svg?branch=master)](https://coveralls.io/github/HotelsDotCom/shunting-yard?branch=master) ![GitHub license](https://img.shields.io/github/license/HotelsDotCom/shunting-yard.svg)
+2. Download and install the latest version of [Circus Train](http://mvnrepository.com/artifact/com.hotels/circus-train/) and set the `CIRCUS_TRAIN_HOME` environment variable:
```
Comment (massdosage) [inline README.md]: Once we sort out the "Install" section for Circus Train let's rather link to the section in its README instead of duplicating links to Maven repos etc. here.

File: README.md [hunk 12]
```diff
@@ -1,61 +1,147 @@
...
+
+         export CIRCUS_TRAIN_HOME=/home/hadoop/circus-train-<circus-train-version>
```
Comment (massdosage) [inline README.md]: Should we use `/opt` as a base instead of `/home/hadoop` in the examples?

File: README.md [hunk 13]
```diff
@@ -1,61 +1,147 @@
...
+## Configuring Shunting Yard
 
-    sudo restart hive-hcatalog-server
+The examples below all demonstrate configuration using YAML and provide fragments covering the most common use cases that should be useful as a basis for building your own configuration. A full configuration reference is provided in the following sections.
 
-### How to install the receiver - TARGET
+### Configuring source, replica and SQS queue
 
-On the target cluster, download and install the latests version of Circus Train and set the `CIRCUS_TRAIN_HOME` environment variable:
+The YAML fragment below shows some common options for setting up the base source (where data is coming from), replica (where data is going to) and the SQS queue to read hive events from.
```
Comment (massdosage) [inline README.md]: Hive

File: README.md [hunk 14]
```diff
@@ -1,61 +1,147 @@
...
+You can obtain Shunting Yard from Maven Central:
+
+[![Maven Central](https://maven-badges.herokuapp.com/maven-central/com.hotels/shunting-yard/badge.svg?subject=com.hotels:shunting-yard)](https://maven-badges.herokuapp.com/maven-central/com.hotels/shunting-yard) [![Build Status](https://travis-ci.org/HotelsDotCom/shunting-yard.svg?branch=master)](https://travis-ci.org/HotelsDotCom/shunting-yard) [![Coverage Status](https://coveralls.io/repos/github/HotelsDotCom/shunting-yard/badge.svg?branch=master)](https://coveralls.io/github/HotelsDotCom/shunting-yard?branch=master) ![GitHub license](https://img.shields.io/github/license/HotelsDotCom/shunting-yard.svg)
 
 ## System architecture
 
 ![Shunting Yard system diagram.](shunting-yard.png "Selected tables getting replicated by Shunting Yard based on Hive events.")
 
-## Start using
+## Install
```
Comment (massdosage) [inline README.md]: I'd recommend an "Overview" section where you explain in a bit more detail what it does and how it works. It can either be a paragraph or two which basically describes the diagram above in words and/or a list of typical steps that happen (a bit like what @AnanaMJ did for describing Housekeeping as a service). I think it would be good for a user to know this before they move on to installing it and configuring and running it.

File: README.md [hunk 15]
```diff
@@ -1,61 +1,147 @@
...
+The examples below all demonstrate configuration using YAML and provide fragments covering the most common use cases that should be useful as a basis for building your own configuration. A full configuration reference is provided in the following sections.
 
-### How to install the receiver - TARGET
+### Configuring source, replica and SQS queue
 
-On the target cluster, download and install the latests version of Circus Train and set the `CIRCUS_TRAIN_HOME` environment variable:
+The YAML fragment below shows some common options for setting up the base source (where data is coming from), replica (where data is going to) and the SQS queue to read hive events from.
 
-    export CIRCUS_TRAIN_HOME=/home/hadoop/circus-train-<circus-train-version>
+    source-catalog:
```
Comment (massdosage) [inline README.md]: Could we have the absolute bare minimum yaml file that would still actually work here? I think it's pretty much what you have but without the `source-table-filter` and `table-replications` sections? That would then replicate everything right? If this is possible that might be a good starting point (with a couple of sentences explaining this). Then the sections below could state that one can add these extra elements and explain what they do?

File: README.md [hunk 16]
```diff
@@ -1,61 +1,147 @@
...
+        - ...
+    table-replications:
+      ...
 
-Export the location of the Hive libraries:
+### Selecting tables to monitor
 
-    export HIVE_LIB=/usr/lib/hive/lib/
-    export HCAT_LIB=/usr/lib/hive-hcatalog/share/hcatalog/
+The YAML fragment below shows how to select the tables to be monitored by Shunting Yard.
```
Comment (massdosage) [inline README.md]: Spell it out a bit more, something like "Shunting Yard by default will replicate all tables in a source Hive Metastore but the user can choose to limit this to a subset of tables by using a source table filter. For example to only replicate tables names `test_table_1` and `test_table_2` you would configure a `source-table-filter` like so:"

File: README.md [hunk 17]
```diff
@@ -1,61 +1,147 @@
...
+    source-table-filter:
+      table-names:
+        - test_database.test_table_1
+        - test_database.test_table_2
 
-    export SHUNTING_YARD_HOME=/home/hadoop/shuting-yard-<version>
+### Specifying target database & table names
 
-Create a `.yml` file under `$SHUNTING_YARD_HOME/conf/` with the correct settings to talk to the messaging infrastructure. You can find guidelines in the file _shunting-yard-minimal.yml.template_. Once this is done you are all set up to run the service:
+The YAML fragments below shows some common options for specifying the target database and table name for the selected tables.
```
Comment (massdosage) [inline README.md]: Similar to above, spell it out what the feature is and then what exactly the example yaml fragment is doing.

File: README.md [hunk 18]
```diff
@@ -1,61 +1,147 @@
...
+    table-replications:
+      - source-table:
+          database-name: source_database
+          table-name: test_table
+        replica-table:
+          database-name: replica_database
+          table-name: test_table_1    
+          
+#### Change only the target database but the table name remains same as source
+
```
Comment (massdosage) [inline README.md]: etc.

File: README.md [hunk 19]
```diff
@@ -1,61 +1,147 @@
...
+    table-replications:
+      - source-table:
+          database-name: source_database
+          table-name: test_table
+        replica-table:
+          database-name: replica_database 
+
+#### Change only the target table name but the database remains same as source
+
```
Comment (massdosage) [inline README.md]: and here. For this and the one above you probably just need to describe what's going on in the yaml as the header already explains the feature.

File: README.md [hunk 20]
```diff
@@ -1,61 +1,147 @@
...
+        replica-table:
+          table-name: test_table_1
+
+### Shunting Yard configuration reference
+The table below describes all the available configuration values for Shunting Yard.
+
+|Property|Required|Description|
+|:----|:----:|:----|
+|`source-catalog.name`|Yes|A name for the source catalog for events and logging.|
+|`source-catalog.hive-metastore-uris`|No|Fully qualified URI of the source cluster's Hive metastore Thrift service.|
```
Comment (massdosage) [inline README.md]: What happens if you don't provide this? We should probably explain...

File: README.md [hunk 21]
```diff
@@ -1,61 +1,147 @@
...
+|`replica-catalog.name`|Yes|A name for the replica catalog for events and logging.|
+|`replica-catalog.hive-metastore-uris`|Yes|Fully qualified URI of the replica cluster's Hive metastore Thrift service.|
+|`event-receiver.configuration-properties.com.hotels.shunting.yard.event.receiver.sqs.queue`|Yes|Fully qualified URI of the [AWS SQS](https://aws.amazon.com/sqs/) Queue to read the Hive events from.|
```
Comment (massdosage) [inline README.md]: Somewhere we should explain that these are actually events coming from Apiary right? Perhaps in the overview right at the top where things are covered in more depth?

File: README.md [hunk 22]
```diff
@@ -1,61 +1,147 @@
...
+|`event-receiver.configuration-properties.com.hotels.shunting.yard.event.receiver.sqs.wait.time.seconds`|No|Wait time in seconds for which the receiver will poll the SQS queue for a batch of messages. Default is 10 seconds. Read more about long polling with AWS SQS [here](https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-long-polling.html).|
+|`source-table-filter.table-names`|No|A list of tables selected for Shunting Yard replication. Supported format:`database_1.table_1, database_2.table_2`|
```
Comment (massdosage) [inline README.md]: For all the values which are optional we should explain what the behaviour is if they're not set to anything.

File: README.md [hunk 23]
```diff
@@ -1,61 +1,147 @@
...
+|`table-replications[n].source-table.database-name`|No|The name of the database in which the table you wish to replicate is located.|
+|`table-replications[n].source-table.table-name`|No|The name of the table which you wish to replicate.|
+|`table-replications[n].replica-table.database-name`|No|The name of the destination database in which to replicate the table. Defaults to source database name.|
```
Comment (massdosage) [inline README.md]: Yes, here and the next one we explain what happens if not set.

File: README.md [hunk 24]
```diff
@@ -1,61 +1,147 @@
...
+|`table-replications[n].replica-table.table-name`|No|The name of the table at the destination. Defaults to source table name.|
+
+### Configuring Graphite metrics
 
 Graphite configurations can be passed to Shunting Yard using an optional `--ct-config` argument which takes a YAML file and passes it directly to internal Circus Train instance. Refer to the [Circus Train README](https://github.com/HotelsDotCom/circus-train#graphite) for more details.
```
Comment (massdosage) [inline README.md]: Make clear this is a *different* YAML configuration file to the one described above.

File: README.md (hunk 8 above)
Comment (abhimanyugupta07) [reply README.md]: Yeah

File: README.md (hunk 3 above)
Comment (AnanaMJ) [reply README.md]: We talked about this and I'm ok with what's in the document now :)

File: README.md (hunk 20 above)
Comment (massdosage) [reply README.md]: This ^

Comment (abhimanyugupta07) [reply README.md]: oh. That is mandatory. Changed now.

File: README.md (hunk 7 above)
Comment (abhimanyugupta07) [reply README.md]: It is there now :)

File: README.md [hunk 25]
```diff
@@ -1,61 +1,191 @@
 # Shunting Yard
 
-A Spring Boot app that reads serialized Hive MetaStore Events and builds a YAML file with the information provided in the event which is then passed to [Circus Train](https://github.com/HotelsDotCom/circus-train) to perform the replication.
+Shunting Yard reads serialized Hive MetaStore Events from a queue ([AWS SQS](https://aws.amazon.com/sqs/) is currently supported) and replicates the data between two data lakes. It does this by building a YAML file with the information provided in the event which is then passed to [Circus Train](https://github.com/HotelsDotCom/circus-train) to perform the replication.
```
Comment (AnanaMJ) [inline README.md]: MetaStore &rightarrow; Metastore

File: README.md [hunk 26]
```diff
@@ -1,61 +1,191 @@
...
+If you are planning to run Shunting Yard on EMR you will need to set up the EMR classpath by exporting the following environment variables before calling the `bin/replicator.sh` script:
+
+```
+export HCAT_LIB=/usr/lib/hive-hcatalog/share/hcatalog/
+export HIVE_LIB=/usr/lib/hive/lib/
+```
 
-    export HIVE_LIB=/usr/lib/hive/lib/
-    export HCAT_LIB=/usr/lib/hive-hcatalog/share/hcatalog/
+Note that the paths above are correct as of when this document was last updated but may differ across EMR versions. Refer to the [EMR release guide](http://docs.aws.amazon.com/emr/latest/ReleaseGuide/emr-release-components.html) for more up to date information if necessary.
```
Comment (AnanaMJ) [inline README.md]: up to date information &rightarrow; up-to-date information

Up to date is without hyphens when it's used after the noun it describes: "The information is up to date."
When it's before the noun, it's with hyphens: "This is an up-to-date information."
source: https://dictionary.cambridge.org/dictionary/english/up-to-date

File: README.md [hunk 27]
```diff
@@ -1,61 +1,191 @@
...
-Graphite configurations can be passed to Shunting Yard using an optional `--ct-config` argument which takes a YAML file and passes it directly to internal Circus Train instance. Refer to the [Circus Train README](https://github.com/HotelsDotCom/circus-train#graphite) for more details.
+    source-table-filter:
+      table-names:
+        - test_database.test_table_1
+        - test_database.test_table_2
 
-#### Sample ct-config.yml:
+### Specifying target database & table names
+
+Shunting Yard will by default replicate the data into the replica data lake with same replica database name and table name as the source. Sometimes a user might need to change the replica database name or table name or both. The YAML fragments below shows some common options for specifying the replica database and table name for the selected tables.
```
Comment (AnanaMJ) [inline README.md]: with same replica database name and table name &rightarrow; with **the** same replica database name and table name

File: README.md [hunk 28]
```diff
@@ -1,61 +1,191 @@
...
+      - source-table:
+          database-name: source_database
+          table-name: test_table
+        replica-table:
+          database-name: replica_database
+          table-name: test_table_1    
+          
+#### Change only the replica database but the table name remains same as source
+
+In this case, the replica table name is not provided in the `table-replications` and hence, it will be same as source table name.
```
Comment (AnanaMJ) [inline README.md]: and hence &rightarrow; ~and~ hence
It's like saying "and therefore" or "and so"

File: README.md (hunk 27 above)
Comment (AnanaMJ) [reply README.md]: If you want to not have that many "the", you can have 
"Shunting Yard will by default replicate data into a replica data lake with the same replica database name and table name as the source."

File: README.md (hunk 26 above)
Comment (abhimanyugupta07) [reply README.md]: .... closes laptop and goes home ;)
